client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java

```diff
@@ -119,14 +133,49 @@ public RestClientBuilder setRequestConfigCallback(RequestConfigCallback requestC
     }
 
     /**
+     * Sets the base path for every request used by the http client.
+     * <p>
+     * In essence, every request's {@code endpoint} is prefixed by this {@code basePath}. The base path is useful for when Elasticsearch is
+     * behind a proxy that provides a base path; it is not intended for other purposes.
+     *
+     * @throws NullPointerException if {@code basePath} is {@code null}.
+     * @throws IllegalArgumentException if {@code basePath} is empty, only '/', or ends with more than one '/'.
+     */
+    public RestClientBuilder setBasePath(String basePath) {
+        Objects.requireNonNull(basePath, "basePath must not be null");
+        String cleanBasePath = basePath.trim();
+
+        if (cleanBasePath.startsWith("/") == false) {
+            cleanBasePath = "/" + cleanBasePath;
+        }
```

Comment:
I think I recall cases that work perfectly without the leading "/". e.g. https://github.com/elastic/elasticsearch/issues/19314